nemo/core/classes/modelPT.py

```diff
@@ -157,86 +164,79 @@ def register_artifact(self, config_path: str, src: str):
-                if self._is_restore_type_tarfile():
-                    # Get path where the command is executed - the artifacts will be "retrieved" there
-                    # (original .nemo behavior)
-                    cwd = os.getcwd()
-                    try:
-                        # Step into the nemo archive to try and find the file
-                        with tempfile.TemporaryDirectory() as tmpdir:
-                            self._unpack_nemo_file(path2file=_MODEL_RESTORE_PATH, out_folder=tmpdir)
-                            os.chdir(tmpdir)
-                            if os.path.exists(basename_src):
-                                logging.warning(f"Using {os.path.abspath(basename_src)} instead of {src}.")
-                                used_src = basename_src
-
-                                archive_item.path = used_src
-                                archive_item.path_type = model_utils.ArtifactPathType.TAR_PATH
-                            else:
-                                # No further action can be taken, file not found anywhere
-                                raise FileNotFoundError(
-                                    f"Could not find {used_src} inside "
-                                    f"tarfile {_MODEL_RESTORE_PATH} or under local"
-                                )
-                    finally:
-                        # change back working directory
-                        os.chdir(cwd)
-                else:
-                    # No further action can be taken, file not found anywhere
-                    raise FileNotFoundError(f"Could not find {used_src}")
-            else:
-                # Found filepath
-                archive_item.path = used_src
+        if config_path is None or config_path.strip()=="":
+            raise ValueError("config_path for artifact can't be None or empty")
+
+        if config_path in self.artifacts.keys():
+            raise ValueError(f"You tried to register an artifact under config key={config_path} but an artifact for"
+                             f"it has already been registered.")
+
+        # src is a local existing path - register artifact and return exact same path for usage by the model
+        if os.path.exists(src):
+            return_path = os.path.abspath(src)
+        # this is the case when artifact must be retried from the nemo file
+        # we are assuming that the location of the right nemo file is available from _MODEL_RESTORE_PATH
+        elif src.startswith("nemo:"):
+            tmpdir = _CACHE_ROOT + f"_TMPDIRS/{uuid.uuid4().hex}"
+            os.makedirs(tmpdir)
+            # TODO: Write a code to delete this in destructor
+            outfolder = self._unpack_nemo_file(path2file=_MODEL_RESTORE_PATH, out_folder=tmpdir)
+            #print(f" AAA: {os.listdir(tmpdir)}, B: {src}")
+            return_path = os.path.abspath(os.path.join(outfolder, src[5:]))
```

Comment:
In this case ArtifactPathType should be tarfile instead of local path.

Reply:
Also, I think we should avoid extracting file into temp dir for a long time like this, it would be better to enter a tarfile and check whether it contained the value of not, finally deleting the temp dir immediately if it doesn't. Deferring cleanup can cause a lot of issues in slurm envs + pollute the local storage if there is failure to delete files.

Reply:
no, it should be local path because I am going to return and refer to this artifact based on local path.

I added destructor to make sure tmp folders get deleted once model instance is destroyed

Reply:
Destructors are not guaranteed to be called, especially in slurm environments. So jobs which run out of time will just keep adding more data to the cache dir. Thats why I had TAR_PATH to denote that at save time, open a temp tarfile and use the artifact inside it.

Reply:
got rid of global cache